pyintegration/tests/test_dtypes.py

```diff
@@ -0,0 +1,194 @@
+            expected = [0, NULL_LONG, 1, NULL_LONG]
+            j_array = dtypes.int64.array_of(np_array, nan_to_value=0)
+            self.assertIn("[J", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+            pd_series = pd.Series(np_array)
+            j_array = dtypes.int64.array_of(pd_series)
+            self.assertIn("[J", str(type(j_array)))
+            expected = [NULL_LONG, NULL_LONG, 1, NULL_LONG]
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("int array from numpy array"):
+            expected = [NULL_INT, NULL_INT, 1, NULL_INT]
+            j_array = dtypes.int32.array_of(np_array)
+            self.assertIn("[I", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("short array from numpy array"):
+            # not sure if this is a numpy bug
+            expected = [0, 0, 1, 0]
+            j_array = dtypes.short.array_of(np_array)
+            self.assertIn("[S", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("byte array from numpy array"):
+            # same as short
+            expected = [0, 0, 1, 0]
+            j_array = dtypes.byte.array_of(np_array)
+            self.assertIn("[B", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("int array from Python list"):
+            expected = [1, 2, 3]
+            j_array = dtypes.int32.array_of([1.1, 2.2, 3.3])
+            self.assertIn("[I", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("short array from Python list with null_to_value"):
+            py_list = [NULL_SHORT, 1000, 2000, 3000]
+            expected = [0, 1000, 2000, 3000]
+            j_array = dtypes.short.array_of(py_list, null_to_value=0)
+            self.assertIn("[S", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("byte array from Python list, down cast"):
+            expected = [1000, 2000, 3000]
+            j_array = dtypes.byte.array_of(expected)
+            self.assertIn("[B", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertNotEqual(expected, py_array)
+
+    def test_floating_array_of(self):
+        with self.subTest("double array from numpy array with nan2value null2value"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            expected = [NULL_DOUBLE, 0, 1.1, float('inf')]
+            j_array = dtypes.float64.array_of(np_array, nan_to_value=NULL_DOUBLE, null_to_value=0.0)
```

Comment:
This is an interesting use case.  I think when reading the conversions that the conversion is done in steps.  If this is the case, you could remap one value and then the second step remaps that value to another default, which is different than you would get by mapping both values at once.